### ISSUE TITLE
libs/common: Temporary File Name Collision

### ISSUE THREAD

**Issue by Liamsi** [tendermint/tendermint#2109]
https://github.com/tendermint/tendermint/blob/40d6dc2ee55a34a075e7b7e73c3525c84ae705ea/libs/common/os.go#L129-L136

> Although unlikely, RandStr() may return the same value during two concurrent executions of WriteFileAtomic(), causing the same temporary file name to be used. This may corrupt the contents within the temporary file.

Recommendation:
> use Go's `TempFile` function to generate a temporary file

**Comment by xla** [comment]
This used to be the case, but has been changed for reasons which escape me now. :>

**Comment by Liamsi** [comment]
I do not thing this is a serious issue but we are probably safer (going back to) using Go's `TempFile`.

**Comment by xla** [comment]
Agreed.

**Comment by ValarDragon** [comment]
I read through the tempfile source. Basically it does a better job here then we will do. https://golang.org/src/io/ioutil/tempfile.go

TL;DR, uses a super fast Linear congruential generator. (Faster than math.rand)
On collision, tries next LCG output.
If 10 collisions in a row, reseeds LCG. It does all this while being thread safe.

**Comment by ValarDragon** [comment]
Oh found the probable reason why, we use the following flags:
`flag = os.O_WRONLY | os.O_CREATE | os.O_SYNC | os.O_TRUNC`
golang's tmp file uses the following:
`os.O_RDWR|os.O_CREATE|os.O_EXCL`.

We want our write only flag, so we should probably the stdlib's logic here.

**Comment by xla** [comment]
@ValarDragon That was it. Thanks for surfacing my own historical decisions. :>
